src/pseudopeople/noise_functions.py

```diff
@@ -149,13 +153,144 @@ def duplicate_with_guardian(
+    # noise level
+
+    noised_data = []
+    for group, group_df in formatted_group_data.items():
+        # Get index groups that can be noised based on dependent and guardian(s) addresses
+        both_different_index = group_df.index[
+            (group_df["household_id"] != group_df["guardian_1_household_id"])
+            & (group_df["guardian_1_household_id"].notna())
+            & (group_df["household_id"] != group_df["guardian_2_household_id"])
+            & (group_df["guardian_2_household_id"].notna())
+        ]
+        # Choose which guardian to copy when dependent lives in different address from both guardains
+        choices = randomness_stream.choice(
+            both_different_index,
+            choices=["guardian_1", "guardian_2"],
+            additional_key=f"{dataset_name}_duplicate_with_guardian_{group}_guardian_choice",
+        )
+        group_df.loc[both_different_index, "copy_guardian"] = choices
+        # Get remaining dependents that live in different address from one of their guardians
+        guardian_1_different_index = group_df.index[
+            (group_df["household_id"] != group_df["guardian_1_household_id"])
+            & (group_df["guardian_1_household_id"].notna())
+        ].difference(choices.index)
+        group_df.loc[guardian_1_different_index, "copy_guardian"] = "guardian_1"
+        guardian_2_different_index = group_df.index[
+            (group_df["household_id"] != group_df["guardian_2_household_id"])
+            & (group_df["guardian_2_household_id"].notna())
+        ].difference(choices.index)
+        group_df.loc[guardian_2_different_index, "copy_guardian"] = "guardian_2"
+
+        # Noise data
+        to_noise_index = get_index_to_noise(
+            group_df,
+            configuration[group],
+            randomness_stream,
+            f"{dataset_name}_duplicate_with_guardian_{group}",
+        )
+        # Copy over address information from guardian to dependent
+        for guardian in ["guardian_1", "guardian_2"]:
+            index_to_copy = to_noise_index.intersection(
+                group_df.index[group_df["copy_guardian"] == guardian]
+            )
+            # Don't try to copy if there are no rows to copy
+            if index_to_copy.empty:
+                continue
+            noised_group_df = group_df.loc[index_to_copy]
+            noised_group_df[GUARDIAN_DUPLICATION_ADDRESS_COLUMNS] = group_df.loc[
+                index_to_copy,
+                [f"{guardian}_" + column for column in GUARDIAN_DUPLICATION_ADDRESS_COLUMNS],
+            ]
+            noised_data.append(noised_group_df)
+
+    # Combine all noised dataframes
+    if not noised_data:
+        return dataset_data
+    else:
+        noised_data = pd.concat(noised_data)
+        noised_data["relationship_to_reference_person"] = "Other relative"
+        # Clense columns
+        noised_data = noised_data.drop(
```

Comment:
nit: A bit more readable perhaps to `noised_data = noised_data[dataset_data.columns]`